File: Java/src/Decoder.java
/**
 * Created by Oliver Poole(12022846) on 25/11/15.
 */
public class Decoder {

    public static String decode(FrequencyTable frequencyTable, double encodedValue) {

        /**
         *  Decoding the output string
         */

        String outputString = "";

        for (int x = 0; x < frequencyTable.numberOfElements; x++) {

            for (Symbol symbol : frequencyTable.table) {
                if ((encodedValue > symbol.lowerRange || (int)encodedValue == (int)symbol.lowerRange)
                        && (encodedValue < symbol.upperRange)) {

                    outputString += symbol.character;

                    encodedValue = (encodedValue - symbol.lowerRange) / (symbol.upperRange - symbol.lowerRange);

                    break;
                }
            }
        }

        return outputString;
    }
}

File: Swift/Arithmetic Coding/Decoder.swift
//
//  Decoder.swift
//  Arithmetic Coding
//
//  Created by Oliver Poole on 24/11/2015.
//  Copyright © 2015 OliverPoole. All rights reserved.
//

import Foundation

class Decoder {
    
    /**
     Decodes the encoded double value
     
     - parameter frequencyTable: The table containing the frequencies of each character
     - parameter encodedValue:   The encoded value ready to be decoded
     
     - returns: A string containing the decoded value
     */
    static func decode(frequencyTable : FrequencyTable, var encodedValue : Double) -> String {
        
        var outputString = ""
        
        // Loop once for each character in the frequency table
        for _ in 1...frequencyTable.numberOfElements {
            
            // Find the symbol where the encoded value currently lies
            let symbol = Array(frequencyTable.table.values).findClosestSymbolTo(encodedValue)!
        
            // Append the symbol to the output string
            outputString += String(symbol.character)
            
            // Remove the symbol from the encoded value
            encodedValue = (encodedValue - symbol.range!.lower) / (symbol.range!.upper - symbol.range!.lower)
            
        }
        
        return outputString
    }
}